core/src/main/java/io/undertow/server/protocol/framed/AbstractFramedStreamSourceChannel.java

```diff
@@ -342,43 +342,43 @@ protected boolean isLastFrame() {
 
     @Override
     public void awaitReadable() throws IOException {
-        if(Thread.currentThread() == getIoThread()) {
+        if (Thread.currentThread() == getIoThread()) {
             throw UndertowMessages.MESSAGES.awaitCalledFromIoThread();
         }
-        if (data == null && pendingFrameData.isEmpty()) {
-            synchronized (lock) {
-                if (data == null && pendingFrameData.isEmpty()) {
-                    try {
-                        waiters++;
-                        lock.wait();
-                    } catch (InterruptedException e) {
-                        Thread.currentThread().interrupt();
-                        throw new InterruptedIOException();
-                    } finally {
-                        waiters--;
-                    }
+        synchronized (lock) {
+            while (data == null && pendingFrameData.isEmpty()) {
+                if (anyAreSet(state, STATE_CLOSED | STATE_STREAM_BROKEN)) return;
+
+                waiters++;
+                try {
+                    lock.wait();
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                    throw new InterruptedIOException();
+                } finally {
+                    waiters--;
                 }
             }
         }
     }
 
     @Override
     public void awaitReadable(long l, TimeUnit timeUnit) throws IOException {
-        if(Thread.currentThread() == getIoThread()) {
+        if (Thread.currentThread() == getIoThread()) {
             throw UndertowMessages.MESSAGES.awaitCalledFromIoThread();
         }
-        if (data == null) {
-            synchronized (lock) {
-                if (data == null) {
-                    try {
-                        waiters++;
-                        lock.wait(timeUnit.toMillis(l));
-                    } catch (InterruptedException e) {
-                        Thread.currentThread().interrupt();
-                        throw new InterruptedIOException();
-                    } finally {
-                        waiters--;
-                    }
+        synchronized (lock) {
+            while (data == null && pendingFrameData.isEmpty()) {
```

Comment:
I don't think this is correct, this may result in waiting longer than the intended duration

Reply:
Thanks for spotting that out - fixed.